### PR TITLE
tests: Add amount compression/decompression fuzzing to existing fuzzing harness

### PR DESCRIPTION
**Pull request by practicalswift**
Small fuzzing improvement:

Add amount compression/decompression fuzzing to existing fuzzing harness: test compression round-trip (`DecompressAmount(CompressAmount(…))`).

Make the domain of `CompressAmount(…)` explicit.

Amount compression primer:

```
    Compact serialization for amounts

    Special serializer/deserializer for amount values. It is optimized for
    values which have few non-zero digits in decimal representation. Most
    amounts currently in the txout set take only 1 or 2 bytes to
    represent.
```

**How to test this PR**

```
$ make distclean
$ ./autogen.sh
$ CC=clang CXX=clang++ ./configure --enable-fuzz \
      --with-sanitizers=address,fuzzer,undefined
$ make
$ src/test/fuzz/integer
…
```

### CODE REVIEW

File: src/compressor.h [hunk 1]
```diff
@@ -19,6 +19,7 @@ bool CompressScript(const CScript& script, std::vector<unsigned char> &out);
 unsigned int GetSpecialScriptSize(unsigned int nSize);
 bool DecompressScript(CScript& script, unsigned int nSize, const std::vector<unsigned char> &out);
 
+// Defined for nAmount <= MAX_MONEY.
```
Comment (MarcoFalke) [inline src/compressor.h]: Do you have any evidence to support this claim?

Comment (kiminuo) [reply src/compressor.h]: The algorithm works for larger numbers fine. Maybe just reword the comment a bit?

https://ideone.com/6TkUTp for tinkering

Comment (sipa) [reply src/compressor.h]: I believe the algorithm works fine up to approximately 2^64 / 9.

Comment (practicalswift) [reply src/compressor.h]: Yes, it is the hard upper limit at `2^64 / 9` that I noticed: above that we wrap around :)

Comment (practicalswift) [reply src/compressor.h]: @kiminuo The point of the comment is to make it clear that this function is not intended to be used with input values above `MAX_MONEY` :)

File: src/test/fuzz/integer.cpp [hunk 2]
```diff
@@ -56,7 +57,11 @@ void test_one_input(const std::vector<uint8_t>& buffer)
 
     const Consensus::Params& consensus_params = Params().GetConsensus();
     (void)CheckProofOfWork(u256, u32, consensus_params);
-    (void)CompressAmount(u64);
+    if (u64 <= MAX_MONEY) {
+        assert(u64 == DecompressAmount(CompressAmount(u64)));
+    } else {
+        (void)CompressAmount(u64);
+    }
```
Comment (MarcoFalke) [inline src/test/fuzz/integer.cpp]: Could add `DecompressAmount(u64);`?

**Review by maflcko — APPROVED** [review]
ACK on the fuzzer. Not sure about the value of the comment.

File: src/test/fuzz/integer.cpp (hunk 2 above)
Comment (practicalswift) [reply src/test/fuzz/integer.cpp]: That is already done on L65 below :)

**Comment by practicalswift** [comment]
@MarcoFalke Do you have any suggestion on how to phrase the precondition comment to make it more valuable?

**Comment by practicalswift** [comment]
@MarcoFalke 

This is an example of a gotcha where the upper limit kicks in, an unsigned integer wraparound occurs and the `CAmount` roundtrip fails:

```
int main(void) {
  using CAmount = int64_t;
  CAmount before = -1;
  CAmount after = DecompressAmount(CompressAmount(before));

  std::cout << "before=" << before << std::endl;
  std::cout << "after=" << after << std::endl;
  assert(before == after);
}
```

`-1` (`int64_t`) gets implicitly converted to `std::numeric_limits<uint64_t>::max()` which is above the upper limit around ~2^64 / 9.

Output:

```
comp.cpp:18:32: runtime error: unsigned integer overflow: 16602069666338596453 * 10 cannot be represented in type 'unsigned long'
comp.cpp:43:7: runtime error: unsigned integer overflow: 2049638230412172401 * 10 cannot be represented in type 'unsigned long'
before=-1
after=2049638230412172324
comp: comp.cpp:59: int main(): Assertion `before == after' failed.
Aborted (core dumped)
```

File: src/compressor.h (hunk 1 above)
Comment (kiminuo) [reply src/compressor.h]: @practicalswift I know you know. Not sure if others know. :) (Ideone snippet was for me to understand the algorithm better.)

File: src/compressor.h [hunk 3]
```diff
@@ -19,7 +19,13 @@ bool CompressScript(const CScript& script, std::vector<unsigned char> &out);
 unsigned int GetSpecialScriptSize(unsigned int nSize);
 bool DecompressScript(CScript& script, unsigned int nSize, const std::vector<unsigned char> &out);
 
+// Defined for 0 <= nAmount <= MAX_MONEY.
```
Comment (laanwj) [inline src/compressor.h]: Please use a doxygen compatible comment for documenting functions, see https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-doxygen-compatible-comments

Comment (practicalswift) [reply src/compressor.h]: Good point. Fixed! Thanks!

**Comment by practicalswift** [comment]
Rebased! :)

File: src/compressor.h [hunk 4]
```diff
@@ -19,7 +19,17 @@ bool CompressScript(const CScript& script, std::vector<unsigned char> &out);
 unsigned int GetSpecialScriptSize(unsigned int nSize);
 bool DecompressScript(CScript& script, unsigned int nSize, const std::vector<unsigned char> &out);
 
+/**
+ * Compress amount.
+ *
+ * Please note that nAmount is of type uint64_t and thus cannot be negative.
```
Comment (MarcoFalke) [inline src/compressor.h]:
```suggestion
 * nAmount is of type uint64_t and thus cannot be negative.
```

"Please note that" is an empty phrase and can be removed.

**Review by maflcko — APPROVED** [review]
ACK. LGTM, just one nit. Let me know if you want to address it or just have this merged as is.

File: src/compressor.h (hunk 4 above)
Comment (practicalswift) [reply src/compressor.h]: Good point! Thanks! Now fixed.

**Pull request merged**